RecoTracker/DeDx/plugins/DeDxHitCalibrator.cc

```diff
@@ -0,0 +1,399 @@
+}
+
+void DeDxHitCalibrator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& tracks = iEvent.getHandle(tracksToken_);
+  const auto& dedxHitInfo = iEvent.get(dedxHitInfoToken_);
+  pixelCalib_.setESObjects(iSetup);
+
+  // creates the output collection
+  auto pixelTrackDeDxHitAss = std::make_unique<reco::TrackDeDxHitsCollection>(reco::TrackRefProd(tracks));
+  auto stripTrackDeDxHitAss = std::make_unique<reco::TrackDeDxHitsCollection>(reco::TrackRefProd(tracks));
+
+  for (size_t i = 0; i < tracks->size(); i++) {
+    const auto& track = reco::TrackRef(tracks, i);
+    const auto& dedxHits = dedxHitInfo[track];
+    if (dedxHits.isNull())
+      continue;
+    reco::DeDxHitCollection pixelHits, stripHits;
+    processHitInfo(*dedxHits, track->p(), pixelHits, stripHits);
+    pixelTrackDeDxHitAss->setValue(i, pixelHits);
+    stripTrackDeDxHitAss->setValue(i, stripHits);
+  }
+
+  iEvent.put(std::move(pixelTrackDeDxHitAss), "PixelHits");
+  iEvent.put(std::move(stripTrackDeDxHitAss), "StripHits");
+}
+
+/*****************************************************************************/
+void DeDxHitCalibrator::processHitInfo(const reco::DeDxHitInfo& info,
+                                       const float& trackMomentum,
+                                       reco::DeDxHitCollection& pixelHits,
+                                       reco::DeDxHitCollection& stripHits) {
+  for (size_t i = 0; i < info.size(); i++) {
+    // Require hits to be complete and compatible
+    const auto& type = info.type(i);
+    if (!(type & (1 << reco::DeDxHitInfo::Complete)) || !(type & (1 << reco::DeDxHitInfo::Compatible)))
+      continue;
+
+    // Effective path length
+    const auto& pl = info.pathlength(i);
+    const auto pathLength = pl * (1. + a_ * std::log(pl / l0_));
+    const DetId& detId = info.detId(i);
+
+    // Strip
+    if (const auto& stripCluster = info.stripCluster(i)) {
+      const auto& thickness =
+          dynamic_cast<const StripGeomDetUnit*>(tkGeom_->idToDet(detId))->surface().bounds().thickness();
+      const auto& det = getDetId(detId, thickness);
+      const auto& chipId = ChipId(detId, stripCluster->barycenter() / 128.);
+      // Fill measured strip deposits
+      const auto& thr = dedxCalib_->thr[det];
+      std::vector<std::pair<double, int> > b;
+      b.reserve(stripCluster->amplitudes().size() + 2);
+      b.emplace_back(thr, isBelow);
+      for (const auto& adc : stripCluster->amplitudes()) {
+        if (adc > 253)
+          b.emplace_back(254., isOver);
+        else
+          b.emplace_back(adc + 0.5, isNormal);
+      }
+      b.emplace_back(thr, isBelow);
+      // Fit
+      const auto& result = fitStripCluster(b, dedxCalib_->alpha[det], 1. / dedxCalib_->sigma[det]);
+      // ADC -> e- -> MeV
+      const auto& energy = correctEnergy(result.first * 217 * MeVPerElectron_, chipId);
+      const auto& esigma = correctEnergy(result.second * 217 * MeVPerElectron_, chipId);
```

Comment:
what's `217` ?

Reply:
This is electronPerAdcDec	(Conversion factor or gain for the electronics, when Deconvolution mode)

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideStripDigitization#:~:text=when%20Peak%20mode-,electronPerAdcDec,-217.0

Reply:
According to current configuration is a different number:

https://github.com/cms-sw/cmssw/blob/1234e950f3ee35b6f39abdeba60b2d1a53c0c891/SimGeneral/MixingModule/python/SiStripSimParameters_cfi.py#L98

Reply:
I see that in https://github.com/cms-sw/cmssw/blob/master/RecoTracker/DeDx/python/dedxEstimators_cff.py#L13 , it is used instead: 3.61e-06 * 265 = 9.5665e-4 , which is also stored in https://github.com/cms-sw/cmssw/blob/master/DataFormats/SiStripCommon/interface/ConstantsForHardwareSystems.h#L47C26-L47C40

So I am thinking of replacing with:

`
      const auto& energy = correctEnergy(result.first * sistrip::MeVperADCStrip, chipId);
      const auto& esigma = correctEnergy(result.second * sistrip::MeVperADCStrip, chipId);
`

Reply:
Yes, ok.

Reply:
Implemented in 9a2b9c087bfb8c4403ff1b8ccc3e37250ac20978